packages/SwingSet/src/kernel/kernel.js

```diff
@@ -660,42 +773,108 @@ export default function buildKernel(
         haveOptionsForVat = true;
       }
     }
-    if (haveOptionsForVat) {
-      /** @type { KernelDeliveryChangeVatOptions } */
-      const kd = harden(['changeVatOptions', optionsForVat]);
-      const vd = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd);
-      const status = await deliverAndLogToVat(vatID, kd, vd);
-      return { ...status, discardFailedDelivery: true };
-    } else {
-      return { discardFailedDelivery: true };
+    if (!haveOptionsForVat) {
+      return NO_DELIVERY_CRANK_RESULTS;
+    }
+
+    /** @type { KernelDeliveryChangeVatOptions } */
+    const kd = harden(['changeVatOptions', optionsForVat]);
+    const vd = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd);
+    const status = await deliverAndLogToVat(vatID, kd, vd);
+    const results = deliveryCrankResults(vatID, status, false); // no meter
+    return harden({ ...results, popDelivery: true });
+  }
+
+  function addComputrons(computrons1, computrons2) {
+    if (computrons1 !== undefined) {
+      assert.typeof(computrons1, 'bigint');
+    }
+    if (computrons2 !== undefined) {
+      assert.typeof(computrons2, 'bigint');
     }
+    // leave undefined if both deliveries lacked numbers
+    if (computrons1 !== undefined || computrons2 !== undefined) {
+      return (computrons1 || 0n) + (computrons2 || 0n);
+    }
+    return undefined;
   }
 
   /**
    *
    * @param {RunQueueEventUpgradeVat} message
-   * @returns {Promise<DeliveryStatus | null>}
+   * @returns {Promise<CrankResults>}
    */
   async function processUpgradeVat(message) {
     assert(vatAdminRootKref, `initializeKernel did not set vatAdminRootKref`);
     const { vatID, upgradeID, bundleID, vatParameters } = message;
     insistCapData(vatParameters);
 
-    if (!vatWarehouse.lookup(vatID)) {
-      return null; // vat is dead, splat
+    const vatInfo = vatWarehouse.lookup(vatID);
+    if (!vatInfo) {
+      return NO_DELIVERY_CRANK_RESULTS; // vat terminated already
     }
+    const { meterID } = vatInfo;
     const vatKeeper = kernelKeeper.provideVatKeeper(vatID);
     /** @type { import('../types-external.js').KernelDeliveryStopVat } */
     const kd1 = harden(['stopVat']);
     const vd1 = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd1);
     const status1 = await deliverAndLogToVat(vatID, kd1, vd1);
-    if (status1.terminate) {
-      // TODO: if stopVat fails, stop now, arrange for everything to
-      // be unwound. TODO: we need to notify caller about the failure
-      console.log(`-- upgrade-vat stopVat failed: ${status1.terminate}`);
+
+    function makeFailure(_errorCD) {
+      insistCapData(_errorCD); // capdata(Error)
+      // const args = [upgradeID, false, JSON.parse(errorCD.body)];
+      // actually we shouldn't reveal the details, so instead we do:
+      const error = {
+        '@qclass': 'error',
+        name: 'Error',
+        message: 'vat-upgrade failure',
+      };
+      const args = [upgradeID, false, error];
```

Comment:
yeah, that'd be nice, I definitely had errors during development because I fed the wrong arguments into this callback, lemme see what I can do quickly

Reply:
Hrm, so `vatUpgradeCallback` runs in userspace, so e.g. it expects `error` to either be `undefined` or a real JS `Error` object. Whereas the `kernel.js` function is assembling capdata that will be deserialized before `vatUpgradeCallback` is invoked. I don't think we have a way to express a type that is the capdata serialization of a different type.

I'll add a type definition to `vatUpgradeCallback`, and I'll add a pointer to that function from the kernel.js code that builds the arguments, but I don't see a way to do more than that right now.

Reply:
Oh we should probably have a helper type that does tuple type -> capdata, but yeah, that's another story.